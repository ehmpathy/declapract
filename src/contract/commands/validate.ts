import { Command, flags } from '@oclif/command';
import { validate } from '../../logic/commands/validate';

export default class Validate extends Command {
  public static description =
    "validate the declared practices, use cases, and examples; checks that these declarations are usable and don't contain declaration errors";

  public static flags = {
    help: flags.help({ char: 'h' }),
    config: flags.string({
      char: 'c',
      description: 'path to the declapract declarations config yml',
      required: true,
      default: 'declapract.declare.yml',
    }),
  };

  public async run() {
    const { flags } = this.parse(Validate);
    const config = flags.config!;

    // generate the code
    const configPath = config.slice(0, 1) === '/' ? config : `${process.cwd()}/${config}`; // if starts with /, consider it as an absolute path
    await validate({
      declarePracticesConfigPath: configPath,
    });
  }
}
