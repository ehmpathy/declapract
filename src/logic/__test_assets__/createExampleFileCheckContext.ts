import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
    required: true,
    getProjectRootDirectory: () => '/some/project/root',
  });
