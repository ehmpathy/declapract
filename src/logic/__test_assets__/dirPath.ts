export const testAssetsDirectoryPath = __dirname;
