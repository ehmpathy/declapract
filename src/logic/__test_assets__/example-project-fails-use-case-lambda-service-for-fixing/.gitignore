# all contents of dir are ignored since tests provision and fix the contents in here
*
!.gitignore
!declapract.use.yml

