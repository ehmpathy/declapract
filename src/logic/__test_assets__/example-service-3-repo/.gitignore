.declapract
node_modules
package-lock.json
