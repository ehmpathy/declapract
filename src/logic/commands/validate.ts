import { readDeclarePracticesConfig } from '../declaration/readDeclarePracticesConfig';

export const validate = async ({
  declarePracticesConfigPath,
}: {
  declarePracticesConfigPath: string;
}) => {
  // read the usage config
  await readDeclarePracticesConfig({ configPath: declarePracticesConfigPath });

  // check that each use case's example passes it's checks; // TODO
  // await Promise.all(
  //   config.useCases.map(async (useCase) => {
  //     if (!useCase.example) return;
  //     const evaluations = await evaluateProjectAgainstPracticeDeclarations({
  //       practices: useCase.practices,
  //       projectRootDirectory: useCase.example.projectRootDirectory,
  //       projectVariables: mockProjectVariablesProxyObject,
  //     });
  //   }),
  // );

  // if it passes all of the above, then its good to go
  console.log('Declarations have validated successfully ðŸŽ‰'); // tslint:disable-line no-console
};
