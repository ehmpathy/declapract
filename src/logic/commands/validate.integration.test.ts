import { testAssetsDirectoryPath } from '../__test_assets__/dirPath';
import { validate } from './validate';

const logSpy = jest.spyOn(console, 'log');

describe('plan', () => {
  beforeEach(() => jest.clearAllMocks());
  it('should be able to plan for an example declarations directory', async () => {
    await validate({
      declarePracticesConfigPath: `${testAssetsDirectoryPath}/example-best-practices-repo/declapract.declare.yml`,
    });
    expect(logSpy.mock.calls).toMatchSnapshot();
  });
});
