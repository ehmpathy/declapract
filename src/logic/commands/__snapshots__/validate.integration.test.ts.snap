// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plan should be able to plan for an example declarations directory 1`] = `
Array [
  Array [
    "Declarations have validated successfully ðŸŽ‰",
  ],
]
`;
